packages/core/src/components/panel-stack/panelStack.tsx

```diff
@@ -0,0 +1,142 @@
+    title: string;
+}
+
+export interface IPanelStackProps extends IProps {
+    /** The panel to show when initially rendering the component */
+    initialPanel: IPanel;
+    /**
+     * Callback invoked when the user presses the back button or a panel wants to close the visible panel.
+     * @param oldPanel The panel that was just removed from the stack.
+     */
+    onClose?(oldPanel: IPanel): void;
+    /**
+     * Callback invoked when a panel wants to open a new panel.
+     * @param newPanel The panel that was just added to the stack.
+     */
+    onOpen?(newPanel: IPanel): void;
+}
+
+export interface IPanelStackState {
+    /** Whether the stack is currently animating the push or pop of a panel. */
+    direction: "push" | "pop";
+    /** The current stack of panels. The first panel in the stack will be displayed. */
+    stack: IPanel[];
+}
+
+export class PanelStack extends React.PureComponent<IPanelStackProps, IPanelStackState> {
+    public constructor(props: IPanelStackProps, context?: any) {
+        super(props, context);
+        this.state = {
+            direction: "push",
+            stack: [props.initialPanel],
+        };
+    }
+
+    public render() {
+        const className = classNames(Classes.PANELSTACK, this.props.className);
+        return (
+            <TransitionGroup className={className} component="div">
+                {this.renderCurrentPanel()}
+            </TransitionGroup>
+        );
+    }
+
+    private renderCurrentPanel() {
+        const { stack } = this.state;
+        const [activePanel, previousPanel] = stack;
+        return (
+            <CSSTransition
+                classNames={`${Classes.PANELSTACK_TRANSITION}-${this.state.direction}`}
+                key={stack.length}
+                timeout={{
+                    enter: 400,
+                    exit: 400,
+                }}
+            >
+                <PanelView
+                    key={stack.length}
+                    panelClassName={this.props.className}
+                    previousPanel={previousPanel}
+                    panel={activePanel}
+                    panelProps={this.getPanelProps()}
+                />
+            </CSSTransition>
+        );
+    }
+
+    private getPanelProps(): IPanelProps {
+        const stackSize = this.state.stack.length;
+        return {
+            closePanel: () => this.closePanel(stackSize),
+            openPanel: this.openPanel,
+        };
+    }
+
+    private closePanel = (stackSize: number) => {
+        this.setState(state => {
+            const allowed =
+                // Make sure this function only works for the panel it was issued to,
+                // otherwise rapidly clicking the back link will close multiple panels.
+                state.stack.length === stackSize &&
+                // Prevent the last panel from being closed.
+                state.stack.length > 1;
```

Comment:
this can be refactored to just `if (condition) { return null }`
I initially wrote it this way because I had problems with `setState` typings, but returning `null` should work and allows you to remove the `allowed` variable entirely.